docs/source/optimising.rst

```diff
@@ -0,0 +1,198 @@
+
+      @PETSc.Log.EventDecorator("foo")
+      def do_something_expensive():
+          ...
+
+  If no arguments are passed to ``PETSc.Log.EventDecorator`` then the
+  event name will be the same as the function.
+
+Caveats
+~~~~~~~
+
+* The ``flamegraph.pl`` script assumes by default that the values
+  in the stack traces are sample counts. This means that if you
+  hover over functions in the SVG it will report the count in terms
+  of 'samples' rather than the correct unit of microseconds. A simple
+  fix to this is to include the command line option ``--countname us``
+  when you generate the SVG. For example:
+
+  .. code-block:: bash
+
+    $ ./flamegraph.pl --countname us foo.txt > foo.svg
+
+* If you use PETSc stages in your code these will be ignored in the flame graph.
+
+* If you call ``PETSc.Log.begin()`` as part of your script/package
+  then profiling will not work as expected. This is because this
+  function starts PETSc's default (flat) logging while we need to
+  use nested logging instead.
+
+  This issue can be avoided with the simple guard:
+
+  .. code-block:: python
+  
+    from firedrake.petsc import OptionsManager
+
+    # If the -log_view flag is passed you don't need to call 
+    # PETSc.Log.begin because it is done automatically.
+    if "log_view" in OptionsManager.commandline_options:
+        PETSc.Log.begin()
+
+Common performance issues
+-------------------------
+
+Calling ``solve`` repeatedly
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When solving PDEs, Firedrake uses a PETSc ``SNES`` (nonlinear solver)
+under the hood. Every time the user calls :py:func:`~firedrake.solving.solve`
+a new ``SNES`` is created and used to solve the problem. This is a
+convenient shorthand for scripts that only need to solve a problem
+once, but it is fairly expensive to set up a new ``SNES`` and so
+repeated calls to :py:func:`~firedrake.solving.solve` will introduce
+some overhead.
+
+To get around this problem, users should instead instantiate
+a variational problem (e.g. :py:class:`~.LinearVariationalProblem`)
+and solver (e.g. :py:class:`~.LinearVariationalSolver`) outside of
+the loop body. An example showing how this is done can be found
+in `this demo <https://firedrakeproject.org/demos/DG_advection.py.html>`_.
+
+Other useful tools
+------------------
+
+Here we present a handful of performance analysis tools that users may
+find useful to run with their codes.
+
+py-spy
+~~~~~~
+
+`py-spy <https://github.com/benfred/py-spy>`_ is a great sampling
+profiler that outputs directly to SVG flame graphs. It allows users
+to see the entire stack trace of the program rather than just the
+annotated PETSc events and unlike most Python profilers it can also
+profile native code.
+
+A flame graph for your Firedrake script can be generated from py-spy with:
+
+.. code-block:: bash
+
+   $ py-spy record -o foo.svg --native -- python myscript.py
+
+Unfortunately, py-spy does not work when run in parallel and it will not
+run on MacOS.
```

Comment:
WFM!

Reply:
> py-spy works on Linux, OSX, Windows and FreeBSD, and supports profiling all recent versions of the CPython interpreter (versions 2.3-2.7 and 3.3-3.9).

Hmm. I thought @sv2518 struggled to get it to work because of this. Will correct.